File: R/skcla_mlp.R
#' Neural Network Classifier
#' @title Multi-layer Perceptron Classifier
#' @description Implements classification using a multi-layer perceptron (MLP).
#' Wraps scikit-learn's `MLPClassifier` through `reticulate`.
#' @param attribute Target attribute name for model building.
#' @param slevels List of possible values for classification target.
#' @param hidden_layer_sizes Number of neurons in each hidden layer.
#' @param activation Activation function for hidden layers. One of
#'   `"relu"`, `"identity"`, `"logistic"`, or `"tanh"`.
#' @param solver Optimizer used for training. One of `"adam"`, `"lbfgs"`, or `"sgd"`.
#' @param alpha L2 penalty (regularization term).
#' @param batch_size Size of minibatches for stochastic optimizers. Use `"auto"` or an integer.
#' @param learning_rate_init Initial learning rate used by stochastic solvers.
#' @param max_iter Maximum number of iterations.
#' @param early_stopping Whether to use early stopping.
#' @return A `skcla_mlp` classifier object.
#'
#' @references
#' Bishop, C. M. (1995). Neural Networks for Pattern Recognition.
#' @examples
#' \dontrun{
#' data(iris)
#' clf <- skcla_mlp(
#'   attribute = "Species",
#'   slevels = levels(iris$Species),
#'   hidden_layer_sizes = c(32, 16),
#'   activation = "relu"
#' )
#' clf <- daltoolbox::fit(clf, iris)
#' pred <- predict(clf, iris)
#' table(pred, iris$Species)
#' }
#' @import daltoolbox
#' @export
skcla_mlp <- function(attribute, slevels,
                      hidden_layer_sizes = c(100),
                      activation = c("relu", "identity", "logistic", "tanh"),
                      solver = c("adam", "lbfgs", "sgd"),
                      alpha = 0.0001,
                      batch_size = "auto",
                      learning_rate_init = 0.001,
                      max_iter = 200,
                      early_stopping = FALSE) {
  activation <- match.arg(activation)
  solver <- match.arg(solver)

  obj <- classification(attribute, slevels)
  cobj <- class(obj)
  objex <- list(
    hidden_layer_sizes = as.integer(hidden_layer_sizes),
    activation = activation,
    solver = solver,
    alpha = as.numeric(alpha),
    batch_size = batch_size,
    learning_rate_init = as.numeric(learning_rate_init),
    max_iter = as.integer(max_iter),
    early_stopping = early_stopping
  )

  obj <- c(obj, objex)
  class(obj) <- c("skcla_mlp", cobj)
  obj
}

#' @import daltoolbox
#' @import reticulate
#' @exportS3Method fit skcla_mlp
fit.skcla_mlp <- function(obj, data, ...) {
  python_path <- system.file("python/skcla_mlp.py", package = "daltoolboxdp")
  if (!file.exists(python_path)) {
    stop("Python source file not found. Please check package installation.")
  }
  reticulate::source_python(python_path)

  if (is.null(obj$model)) {
    obj$model <- skcla_mlp_create(
      hidden_layer_sizes = obj$hidden_layer_sizes,
      activation = obj$activation,
      solver = obj$solver,
      alpha = obj$alpha,
      batch_size = obj$batch_size,
      learning_rate_init = obj$learning_rate_init,
      max_iter = obj$max_iter,
      early_stopping = obj$early_stopping
    )
  }

  prepared <- prepare_skcla_fit(obj, data)
  obj <- prepared$obj
  data <- prepared$data
  obj$model <- skcla_mlp_fit(obj$model, data, obj$attribute)

  obj
}

#' @import daltoolbox
#' @import reticulate
#' @export
predict.skcla_mlp <- function(object, x, ...) {
  if (!exists("skcla_mlp_predict_proba")) {
    python_path <- system.file("python/skcla_mlp.py", package = "daltoolboxdp")
    if (!file.exists(python_path)) {
      stop("Python source file not found. Please check package installation.")
    }
    reticulate::source_python(python_path)
  }

  x <- prepare_skcla_predict_data(object, x)

  prediction <- skcla_mlp_predict_proba(object$model, x)
  prediction <- skcla_as_probability(prediction, object$slevels, object$model$classes_)

  prediction
}

File: inst/python/skcla_mlp.py
"""
MLPClassifier wrapper used by daltoolboxdp via reticulate.

R entry points (see R/skcla_mlp.R):
  - skcla_mlp_create(...hyperparams...) -> sklearn model
  - skcla_mlp_fit(model, df_train, target_column) -> fitted model
  - skcla_mlp_predict(model, df_test) -> list of labels
  - skcla_mlp_predict_proba(model, df_test) -> list of per-class probabilities
"""

from sklearn.neural_network import MLPClassifier
import pandas as pd

def skcla_mlp_create(hidden_layer_sizes=(100,), activation='relu', solver='adam', alpha=0.0001, batch_size='auto',
               learning_rate_init=0.001, max_iter=200, early_stopping=False):

    if activation is None:
        activation = 'relu'  
    
    if solver is None:
        solver = 'adam'
    
    model = MLPClassifier(
        hidden_layer_sizes=hidden_layer_sizes,
        activation=activation,
        solver=solver,
        alpha=alpha,
        batch_size=batch_size,
        learning_rate_init=learning_rate_init,
        max_iter=max_iter,
        early_stopping=early_stopping,
    )
    return model

def skcla_mlp_fit(model, df_train, target_column):
    """Fit MLP. df_train must include the target_column."""
    df_train = pd.DataFrame(df_train)
    X_train = df_train.drop(columns=[target_column])
    y_train = df_train[target_column].values

    model.fit(X_train, y_train)
    return model

def skcla_mlp_predict(model, df_test):
    """Predict labels as a Python list to simplify R interop."""
    try:
        df_test = pd.DataFrame(df_test)
        predictions = model.predict(df_test)
        return predictions.tolist()
    except TypeError as e:
        print(f"Error occurred: {e}")
    except Exception as e:
        print(f"Error occurred: {e}")

def skcla_mlp_predict_proba(model, df_test):
    """Predict class probabilities as a nested list to simplify R interop."""
    try:
        df_test = pd.DataFrame(df_test)
        probabilities = model.predict_proba(df_test)
        return probabilities.tolist()
    except TypeError as e:
        print(f"Error occurred: {e}")
        return []
    except Exception as e:
        print(f"Error occurred: {e}")
        return []
